### PR TITLE
Add model and endpoints handshake

### PR DESCRIPTION
**Pull request by mjoyce91**
- Bureau users with appropriate bureau permissions for a cycle position can offer/revoke a handshake
- CDOs can accept/decline a handshake for any bidder
- Bidders can accept/decline a handshake offered to them
- Responses for bid list and bidder list display handshake offered status from database

### CODE REVIEW

**Comment by mjoyce91** [comment]
I just realized I didn't update the CSV export

File: talentmap_api/fsbid/services/bid.py [hunk 1]
```diff
@@ -94,6 +94,43 @@ def remove_bid(employeeId, cyclePositionId, jwt_token):
...
+            hsStatus = pydash.get(hs, 'status')
+
+            isCDOUpdate = pydash.get(hs, 'is_cdo_update') == 1
+
+            if hsStatus == 'O':
+                clonedBids[idx]['hs_status_code'] = 'handshake_offered'
+            if hsStatus == 'A':
+                clonedBids[idx]['hs_status_code'] = 'handshake_accepted'
+            if hsStatus == 'D':
+                clonedBids[idx]['hs_status_code'] = 'handshake_declined'
```
Comment (elizabeth-jimenez) [inline talentmap_api/fsbid/services/bid.py:119]: we can fake a switch statement here 🦜 

```
hsStatuses = {
       'O': "handshake_offered",
        'A': "handshake_accepted",
        'D': "handshake_declined",
    }
  
 clonedBids[idx]['hs_status_code'] = hsStatuses.get(hsStatus, "not_offered")
```

File: talentmap_api/fsbid/services/bid.py [hunk 2]
```diff
@@ -94,6 +94,43 @@ def remove_bid(employeeId, cyclePositionId, jwt_token):
...
+                clonedBids[idx]['hs_cdo_indicator'] = True
+    return clonedBids
+
+
+def map_bids_handshake_status_by_cp_id(bids, cp_id):
+    return map_bids_handshake_status(bids, {'cp_id': cp_id})
+
+
+def map_bids_handshake_status_by_perdet(bids, perdet):
+    return map_bids_handshake_status(bids, {'bidder_perdet': perdet})
```
Comment (elizabeth-jimenez) [inline talentmap_api/fsbid/services/bid.py:131]: 😻  i like how we are re-using `map_bids_handshake_status` by using these. Stealing 👝

File: talentmap_api/fsbid/services/bid.py (hunk 1 above)
Comment (mjoyce91) [reply talentmap_api/fsbid/services/bid.py:119]: o that's cool

File: talentmap_api/bureau/services/available_bidders.py [hunk 3]
```diff
@@ -57,7 +57,7 @@ def get_available_bidders_csv(request):
     }
 
     for record in data["results"]:
-        languages = f'' if pydash.get(record, ["languages"]) else "None listed"
+        languages = '' if pydash.get(record, ["languages"]) else "None listed"
```
Comment (elizabeth-jimenez) [inline talentmap_api/bureau/services/available_bidders.py:60]: Is this defaulting to an empty string if `record[languages]` returns something? I prolly did this 😭

Comment (mjoyce91) [reply talentmap_api/bureau/services/available_bidders.py:60]: I didn't actually change any functionality. `f` is used for string interpolation, but this is already an empty string so it's not needed

Comment (elizabeth-jimenez) [reply talentmap_api/bureau/services/available_bidders.py:60]: I know :) was just wondering if you could confirm my suspicion; I worked on the BE for AB so I am guilty 🐛

File: talentmap_api/fsbid/services/bid.py [hunk 4]
```diff
@@ -94,6 +94,43 @@ def remove_bid(employeeId, cyclePositionId, jwt_token):
...
+
+            if hsStatus == 'O':
+                clonedBids[idx]['hs_status_code'] = 'handshake_offered'
+            if hsStatus == 'A':
+                clonedBids[idx]['hs_status_code'] = 'handshake_accepted'
+            if hsStatus == 'D':
+                clonedBids[idx]['hs_status_code'] = 'handshake_declined'
+
+            if isCDOUpdate:
+                clonedBids[idx]['hs_cdo_indicator'] = True
```
Comment (elizabeth-jimenez) [inline talentmap_api/fsbid/services/bid.py:122]: We might be able to set this on line 112? 🤔

File: talentmap_api/bidding/views/bidhandshake.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    permission_classes = [Or(isDjangoGroupMember('bureau_user'), ) ]
+
+    def put(self, serializer, pk, cp_id, **ars):
+        '''
+        Offers a handshake to a bidder for a cp_id
+        '''
+        hasBureauPermissions = empservices.has_bureau_permissions(cp_id, self.request.META['HTTP_JWT'])
+
+        if not hasBureauPermissions:
+            raise PermissionDenied()
```
Comment (elizabeth-jimenez) [inline talentmap_api/bidding/views/bidhandshake.py:50]: 🤩

**Comment by elizabeth-jimenez** [comment]
The code looks good! I'll throw some edge cases at Swagger tomorrow and drop them here.

File: talentmap_api/bidding/views/bidhandshake.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+        user = UserProfile.objects.get(user=self.request.user)
+        hs = BidHandshake.objects.filter(bidder_perdet=pk, cp_id=cp_id)
+
+        if hs.exists():
+            hs.update(last_editing_user=user, status='O')
```
Comment (elizabeth-jimenez) [inline talentmap_api/bidding/views/bidhandshake.py:56]: should we add the `update_date=datetime.now()` like we do for revoking?

Comment (mjoyce91) [reply talentmap_api/bidding/views/bidhandshake.py:56]: good catch!

**Comment by elizabeth-jimenez** [comment]
Bureau user with perms
- [x] can offer HS: S=O
- [x] no error on re-offering a HS. **timestamp doesn’t update**
- [x] error on offering a HS a bidder has not bid on: **HS offered, do we want to be able to extend a HS to someone that hasn’t bid on the position?**
- [x] can revoke HS: updates last_editing_user, time stamp and S=R
- [x] no error on re-revoking.
- [x] error on revoking a HS that doesn’t exist: 404

Bureau user without perms(should Error on all)
- [x] cannot offer HS: 403
- [x] cannot revoke HS: 403

CDO
- [x] can accept a non-revoked HS for bidder: is_cdo_update, last_editing_bidder, time updates, S=A
- [x] no error on re-accepting a non-revoked HS for bidder
- [x] cannot accept a revoked HS for bidder: 404
- [x] cannot accept a non-existent HS
- [x] can decline a non-revoked HS for bidder(even if already accepted): S=D
- [x] no error on re-declining a non-revoked HS for bidder
- [x] cannot decline a revoked HS: 404

Bidders
- [x] can accept a non-revoked HS: is_cdo_update, last_editing_bidder, time updates, S=A
- [x] no error on re-accepting a non-revoked HS
- [x] cannot accept a revoked HS: 404
- [x] cannot accept a non-existent HS
- [x] can decline a non-revoked HS. S=D
- [x] no error on re-declining a non-revoked HS
- [x] cannot decline a revoked HS: 404

File: talentmap_api/bureau/services/available_bidders.py (hunk 3 above)
Comment (mjoyce91) [reply talentmap_api/bureau/services/available_bidders.py:60]: Oh I see what you mean. Yes if `pydash.get(record, ["languages"])` is a truthy value, then empty string is returned

File: talentmap_api/fsbid/services/bid.py (hunk 4 above)
Comment (mjoyce91) [reply talentmap_api/fsbid/services/bid.py:123]: yep you're right

File: talentmap_api/bureau/services/available_bidders.py (hunk 3 above)
Comment (elizabeth-jimenez) [reply talentmap_api/bureau/services/available_bidders.py:60]: i thought it was a 🐛 but later on in the file it handles this properly 👯  but good catch on that `f''`

**Comment by mjoyce91** [comment]
@elizabeth-jimenez - added a code comment https://github.com/MetaPhase-Consulting/State-TalentMAP-API/pull/496/commits/afc17669b62c000b8a84e1e3e5b5d6356ee1f724 to address your question:

> HS offered, do we want to be able to extend a HS to someone that hasn’t bid on the position?

May not tackle this yet for a version 0

File: talentmap_api/bidding/views/bidhandshake.py [hunk 7]
```diff
@@ -44,6 +44,8 @@ def put(self, serializer, pk, cp_id, **ars):
         '''
         Offers a handshake to a bidder for a cp_id
         '''
+        # TODO: should we limit this endpoint to only bidder perdets of those who have actually bid on this cp?
+        # Is it worth the extra network request for the extra validation?
```
Comment (elizabeth-jimenez) [inline talentmap_api/bidding/views/bidhandshake.py:48]: 😭

**Review by elizabeth-jimenez — APPROVED** [review]
🍰 🧁 🍬 👍

**Pull request merged**